lib/enkf/enkf_main.cpp

```diff
@@ -2283,61 +2283,81 @@ int enkf_main_load_from_forward_model(enkf_main_type * enkf_main, int iter , boo
-    arg_pack_type * arg_pack = arg_pack_alloc();
-    arg_list[iens] = arg_pack;
-
-    if (bool_vector_iget(iactive, iens)) {
-      printf("... ");
-      enkf_state_type * enkf_state = enkf_main_iget_state( enkf_main , iens );
-      arg_pack_append_ptr( arg_pack , enkf_state);                                         /* 0: enkf_state*/
-      arg_pack_append_ptr( arg_pack , ert_run_context_iget_arg( run_context , iens ));     /* 1: run_arg */
-      arg_pack_append_ptr(arg_pack, realizations_msg_list[iens]);                          /* 2: List of interactive mode messages. */
-      arg_pack_append_bool( arg_pack, true );                                              /* 3: Manual load */
-      arg_pack_append_ptr(arg_pack, &result[iens]);                                        /* 4: Result */
-      thread_pool_add_job( tp , enkf_state_load_from_forward_model_mt , arg_pack);
-    }
-    printf("done\n");
-  }
-
-  thread_pool_join( tp );
-  thread_pool_free( tp );
-  printf("\n");
-
-  int loaded = 0;
-  for (int iens = 0; iens < ens_size; ++iens) {
-    if (bool_vector_iget(iactive, iens)) {
-      if (result[iens] & LOAD_FAILURE)
-        fprintf(stderr, "** Warning: Function %s: Realization %d load failure\n", __func__, iens);
-      else if (result[iens] & REPORT_STEP_INCOMPATIBLE)
-        fprintf(stderr, "** Warning: Function %s: Realization %d report step incompatible\n", __func__, iens);
-      else
-        loaded++;
-    }
-    arg_pack_free(arg_list[iens]);
-  }
-  free( arg_list );
-  ert_run_context_free( run_context );
+  int loaded = enkf_main_load_from_run_context(enkf_main, run_context, realizations_msg_list, fs);
+  ert_run_context_free(run_context);
   return loaded;
 }
 
 
+int enkf_main_load_from_run_context_from_gui(enkf_main_type* enkf_main, ert_run_context_type* run_context, enkf_fs_type* fs) {
+   auto const ens_size = enkf_main_get_ensemble_size(enkf_main);
+   stringlist_type ** realizations_msg_list = (stringlist_type **) util_calloc(ens_size, sizeof *realizations_msg_list); // CXX_CAST_ERROR
+   for(int iens = 0; iens < ens_size; ++iens)
+      realizations_msg_list[iens] = stringlist_alloc_new();
+
+   int loaded = enkf_main_load_from_run_context(enkf_main, run_context, realizations_msg_list, fs);
+
+   for(int iens = 0; iens < ens_size; ++iens)
+      stringlist_free(realizations_msg_list[iens]);
+   free(realizations_msg_list);
+   return loaded;
+}
+
+int enkf_main_load_from_run_context(
+      enkf_main_type * enkf_main,
+      ert_run_context_type * run_context,
+      stringlist_type ** realizations_msg_list,
+      enkf_fs_type * fs) {
+   auto const ens_size = enkf_main_get_ensemble_size( enkf_main );
+   auto const * iactive = ert_run_context_get_iactive(run_context);
+
+   int result[ens_size];
+   arg_pack_type ** arg_list = (arg_pack_type **) util_calloc( ens_size , sizeof * arg_list ); // CXX_CAST_ERROR
+   thread_pool_type * tp     = thread_pool_alloc( 4 , true );  /* num_cpu - HARD coded. */
```

Comment:
What happens if available cpu is less than 4?

Reply:
Nothing in particular, performance will of course be the best if the number of workers (i.e. the 4 here) and the number of cpu is the same - but correctness is not affected.

The C++11 function `std::thread::hardware_concurrency()` could probably be used instead of the hardcoded 4?